src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/RenameTypeService.cs

```diff
@@ -0,0 +1,189 @@
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS
+{
+    [Export(typeof(IRenameTypeService))]
+    internal class RenameTypeService : IRenameTypeService
+    {
+        private readonly VisualStudioWorkspace _workspace;
+
+        [ImportingConstructor]
+        public RenameTypeService(
+            UnconfiguredProject project,
+            VisualStudioWorkspace visualStudioWorkspace)
+        {
+            _workspace = visualStudioWorkspace;
+            SyntaxFactsServicesImpl = new OrderPrecedenceImportCollection<ISyntaxFactsService>(projectCapabilityCheckProvider: project);
+        }
+
+        [ImportMany]
+        protected OrderPrecedenceImportCollection<ISyntaxFactsService> SyntaxFactsServicesImpl { get; }
+
+        private ISyntaxFactsService SyntaxFactsService => SyntaxFactsServicesImpl.First().Value;
+
+        public async Task<bool> DoesFileNameMatchTypeAsync(string filePath)
+        {
+            string name = Path.GetFileNameWithoutExtension(filePath);
+            ISymbol symbol = await TryGetSymbolToRenameAsync(name, filePath);
+            return symbol != null;
+        }
+
+        private string _expectedFilePath = null;
+        private static readonly object s_filePathLock = new object();
+
+        public void QueueFileWithTypeToBeRenamed(string filePath)
+        {
+            lock (s_filePathLock)
+            {
+                _expectedFilePath = filePath;
+            }
+        }
+
+        public async Task<bool> AnyTypeToRenameAsync(string oldFilePath, string newFilePath)
+        {
+            string expectedFilePath = _expectedFilePath;
+            if (expectedFilePath != oldFilePath)
+                return false;
+
+            string oldName = Path.GetFileNameWithoutExtension(oldFilePath);
+            string newName = Path.GetFileNameWithoutExtension(newFilePath);
+
+            if (!CanHandleRename(oldName, newName))
+                return false;
+
+            ISymbol symbol = await TryGetSymbolToRenameAsync(oldName, newFilePath);
+            return symbol != null;
+        }
+
+        public bool RenameType(string oldFilePath, string newFilePath, CancellationToken cancellationToken)
+        {
+            string expectedFilePath = _expectedFilePath;
+            if (expectedFilePath != oldFilePath)
+                return false;
+
+            string oldName = Path.GetFileNameWithoutExtension(oldFilePath);
+            string newName = Path.GetFileNameWithoutExtension(newFilePath);
+
+            if (!TryGetDocument(newFilePath, out Document document))
+                return false;
+
+            EnvDTE.FileCodeModel codeModel = _workspace.GetFileCodeModel(document.Id);
+
+            if (!TryGetCodeElementToRename(oldName, codeModel.CodeElements, out EnvDTE80.CodeElement2 codeElementToRename))
+                return false;
+
+            codeElementToRename.RenameSymbol(newName);
+            return true;
+        }
+
+        private bool CanHandleRename(string oldName, string newName)
```

Comment:
These 4 methods could be inlined into single method CanHandleRename and it would be smaller than them combined.